test/Base_Tests/src/Runtime/GC_Example.enso

```diff
@@ -15,11 +15,11 @@ type My_Resource
 
```

Comment:
Hmmm that may be true

Reply:
We can trivially fix the test, if we allocate the resources first and only later clear them.

Reply:
How will that help? We will not get clashes between `+1` and `-1` but we can still get problems due to non-atomic `+1` (and `-1` respectively) in each batch.

Reply:
I mean this problem among even uniform increments: https://github.com/openjdk/jcstress/blob/master/jcstress-samples/src/main/java/org/openjdk/jcstress/samples/api/API_01_Simple.java

i.e. if there's `N` actors concurrently incrementing non-atomically, we don't have the guarantee that all `N` increments will be registered - so we will still get the test stuck.

Reply:
What if we do the following:
1. Introduce a timeout - if the cleanup did not succeed after ~5 minutes - let's make that a test failure - this will avoid any potential problem from blocking the runner for several hours, and it will actually make the tests report come up and be more helpful.
2. Maybe add some margins for the test to finish? I.e. do it in phases like you suggest, check that e.g. 100 000 resources caused the counter to increment to at least 90 000, and then check that after cleanup it went below 10 000 (and be OK with it getting below 0).

(1) will ensure we are not stuck and (2) should make it far less likely to fail the test due to concurrent behaviour, while still testing the core of this issue - multiple finalizers are scheduled at once.

Reply:
Solved in 3c93430 by holding all the resources in a `Vector` and releasing them only after 3rd round of `Runtime.gc`.

Reply:
> I mean this problem among even uniform increments: https://github.com/openjdk/jcstress/blob/master/jcstress-samples/src/main/java/org/openjdk/jcstress/samples/api/API_01_Simple.java
> 
> i.e. if there's `N` actors concurrently incrementing non-atomically, we don't have the guarantee that all `N` increments will be registered - so we will still get the test stuck.

Please note that the test is running completely _single threaded_. It just schedules _interrupt_ (e.g. _safepoint action_ with a finalizer) at a random moment. 

If the _finalizer_ runs while we are still allocating resources, then we may have two methods on stack that manipulate the `Ref` `counter` and their code is certainly not ready for re-entrance. However if we first allocate everything and only then start running finalizers, then there always be only a single method on stack manipulating the counter - as we disallow _nesting of finalizer interrupts_  since #10602 and there is no longer any other method manipulating the `counter` on stack.

Unless I am mistaken the failure has always happened in _native image_ compiled `enso` launcher. Probably because the Truffle compilation there happens more quickly than in JVM and the code was still allocating... while in the JVM the compilation happens later when all the resources are already allocated. Since 3c93430 this difference shall be suppressed.

Reply:
Ah, right! Thanks for the explanation, I forgot that we now disallow finalizers from interrupting each other, so indeed there is no place for unexpected races like that. I was wrongly viewing it as multithreaded

Reply:
Why do we need the 3 iterations of `Runtime.gc` before we release the resources?

Intuitively it feels like it should be enough to release the resources right away after they all were created. I don't understand what point these 3 iterations of GC server.

Reply:
Do you want to rewrite the `GC_Example` code? It is your code from #10211 anyway. I am just trying to make it reliable.

Reply:
It feels like there's not that much to change, but sure - I can try.

But I need to understand - was there any reason why we'd call the 3 GC rounds before starting the cleanup? From my perspective there is no reason so I'd change to start GCing only afterwards, but I'm afraid I may be missing something and in such case my rewrite could break stuff again.

Reply:
Also if I understand your explanation correctly - my suggestion with 'margins' is not really needed because as it is running on a single thread, we should always get back to `0` after cleanup is done, right?

So the only meaningful change would be to add the timeout?

Reply:
>  suggestion with 'margins' is not really needed

The whole point of the test was to verify we always get the _exact same number_ of allocations as well as finalizer calls. "margins" just defeat the purpose of the test, I'd say.

Reply:
Yes, I think you are right.

Reply:
I've modified the test to fail if a timeout of 5 minutes is exceeded:
```
Allocating 100000 resources...ocate lots of resources at once
Cleaning up 100000 resources
Still 368 resources to clean up...
Still 368 resources to clean up...
Still 368 resources to clean up...
Still 368 resources to clean up...
Still 368 resources to clean up...
Timeout has been reached, but there are 368 resources remaining that were not cleaned up.
[FAILED] Managed_Resource: [3/4, 60348ms]

    - [FAILED] allocate lots of resources at once [60222ms]
        Reason: An unexpected panic was thrown: (Illegal_State.Error 'Timeout has been reached, but there are 368 resources remaining that were not cleaned up.' Nothing)
        at <enso> Panic.throw(Internal)
```

Reply:
Five minutes timeout is a bit too excessive for something that runs in 5s, but we don't expect the test to fail anymore.

Reply:
Fair, I wanted to have a larger timeout to avoid spurious failures under load, 5 mins is still better than 6 hours. But if you think we should change I'm happy to change it e.g. to 1 min or even 30s.